serde/json: experimental asynchronous streaming parser

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
serde/json: Implement string parsing

Implements parsing of JSON strings, including standard escapes and
unicode escapes.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/35126eff-b120-4b02-9af7-f3be2a0d83da?refs=8256c00f317e062d0d2f87b60c8bad817d4a47fb&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/35126eff-b120-4b02-9af7-f3be2a0d83da/license?refs=8256c00f317e062d0d2f87b60c8bad817d4a47fb&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/a0b62d01-3ad4-4022-ae34-f8af2b2a093f?refs=8256c00f317e062d0d2f87b60c8bad817d4a47fb&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/a0b62d01-3ad4-4022-ae34-f8af2b2a093f/license?refs=8256c00f317e062d0d2f87b60c8bad817d4a47fb&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/2a8a99ab-7954-4ae0-890a-7140ce377af0?refs=8256c00f317e062d0d2f87b60c8bad817d4a47fb&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/2a8a99ab-7954-4ae0-890a-7140ce377af0/license?refs=8256c00f317e062d0d2f87b60c8bad817d4a47fb&source=prChecksComment)

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--66fbe7aa-6c9d-486b-a1de-edc96dcbe7fe-->
<!--TEST RESULTS-->
<!--3eeac4f5-ff94-4a4a-b61e-539bde6f25c3-->
<!--TEST RESULTS-->
<!--5b12b605-4133-48f5-812e-4ec74a03d57e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65426">build#65426</a></summary>


| test_id                                                                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65426#01968e36-91bb-4151-a04a-89f0792f5bd0 | FLAKY         | 15/21    |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65426#01968e49-4fe2-4eb2-8555-d34d355d66ba | FAIL          | 0/1      |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=None.allow_gaps_cluster_level=True                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65426#01968e36-91bc-447a-ac0e-afa5ff963584 | FLAKY         | 20/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65426#01968e49-4fe1-4d99-a59a-a0dcf8ac7cb9 | FLAKY         | 12/21    |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=1                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65426#01968e36-91bd-4185-be09-f6254efc44ab | FLAKY         | 20/21    |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65426#01968e49-4fe1-4d99-a59a-a0dcf8ac7cb9 | FLAKY         | 20/21    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65790">build#65790</a></summary>


| test_class               | test_method                              | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-----------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AvailabilityTests        | test_recovery_after_catastrophic_failure |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65790#0196bbeb-0108-435b-82e8-a8528e9952bc | FLAKY         | 20/21    | upstream reliability is '99.16666666666667'. current run reliability is '95.23809523809523'. drift is 3.92857 and the allowed drift is set to 50. The test should PASS |
| DatalakeTransactionTests | test_with_transactions                   | {"cloud_storage_type": 1, "compaction": false}                                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65790#0196bbeb-0108-4f0d-bc8e-a526beeb58be | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionBalancerTest    | test_fuzz_admin_ops                      |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65790#0196bbe1-1c0f-4d1a-81bc-90be291f327f | FLAKY         | 18/21    | upstream reliability is '91.43968871595331'. current run reliability is '85.71428571428571'. drift is 5.7254 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations                     | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65790#0196bbe1-1c0e-41e0-941b-68d342181163 | FLAKY         | 20/21    | upstream reliability is '98.10426540284361'. current run reliability is '95.23809523809523'. drift is 2.86617 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67813">build#67813</a></summary>


| test_class                    | test_method                               | test_arguments                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------------|:------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67813#0197a32e-2297-458a-b103-d89d064d5d55 | FLAKY         | 12/21    | upstream reliability is '72.1419185282523'. current run reliability is '57.14285714285714'. drift is 14.99906 and the allowed drift is set to 50. The test should PASS |
| NodesDecommissioningTest      | test_decommissioning_working_node         | {"delete_topic": true, "tick_interval": 3600000}                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67813#0197a335-b082-4471-8887-68d5504a778d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| NodesDecommissioningTest      | test_recycle_all_nodes                    | {"auto_assign_node_id": false}                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67813#0197a335-b081-470f-8ffe-01dc5808f377 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| TieredStorageTest             | test_tiered_storage                       | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67813#0197a335-b081-470f-8ffe-01dc5808f377 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
</details>
<!--/body-->

rockwotj commented on src/v/serde/json/detail/string.cc:
```diff
@@ -0,0 +1,249 @@
+namespace {
+
+// RapidJSON escape character table.
+#define Z16 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
```
possible to consteval something instead of macro?

rockwotj commented on src/v/serde/json/detail/string.cc:
```diff
@@ -0,0 +1,249 @@
+#undef Z16
+
+// RapidJSON codepoint conversion.
+static void append_codepoint(iobuf& buf, unsigned codepoint) {
```
it's been a hot second but this is utf8 encoding right?

wdberkeley replied:
Yeah, this routine appends the UTF-8 encoding of `codepoint` to `buf`. It doesn't correctly handle invalid codepoints > 0x10FFFF. The plan is to merge this change and follow up with improvements like more validation of strings and string encodings.

rockwotj commented on src/v/serde/json/detail/tests/string_test.cc:
```diff
@@ -0,0 +1,224 @@
+        return;
+    }
+
+    // Try splitting the input at every possible position.
```
Love it. Was going to suggest testing the power set of split points (all combos of valid splits) but this is probably just as effective and cheaper

rockwotj reviewed (commented):
nice start

nvartolomei committed:
serde/json: Implement number parsing

Implements parsing of JSON number values, including parsing integer and
double values.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Add JSON SAX parser

Implements a coroutine-friendly SAX-style JSON push parser.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Add jsonchecker testdata

Adds examples of valid and invalid JSON that will be used by tests added
in subsequent commits.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

dotnwat reviewed (commented):
this looks like a good candidate as a clang fuzzer target? we have a few examples in the tree.

nvartolomei committed:
serde/json: Add json-test-suite testdata

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Add json-test-suite test

Adds tests that check that the parser can parse json-test-suite data.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Add json-test-suite bench

Adds a benchmark for json-test-suite data, comparing the JSON parser
implementation with rapidjson.

Sample results:

single run iterations:    0
single run duration:      1.000s
number of runs:           5
number of cores:          1
random seed:              990765220

test                         iterations      median         mad         min         max      allocs       tasks        inst      cycles
json_parser_bench.serde              21    47.231ms    23.780us    47.103ms    47.257ms   33520.029  548260.000 720249365.7         0.0
json_parser_bench.rapidjson         219     4.570ms    10.505us     4.559ms     4.662ms       2.000       0.000  94968645.4         0.0

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Add DOM model for testing

Implements a basic DOM for testing.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Test conformance with rapidjson

Implements a test that compares the DOM models parsed by our parser and
rapidjson, ensuring that our parser agrees with rapidjson.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Add DOM model for testing rapidjson

Adds support for parsing JSON using rapidjson and converting it into the
DOM model. This will enable comparison testing between our parser and
rapidjson.

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

nvartolomei committed:
serde/json: Add design doc for streaming JSON parser

Co-authored-by: Will Berkeley <will.berkeley@redpanda.com>

rockwotj commented:
sorry @dotnwat didn't mean to request a review from you :)

rockwotj commented on src/v/serde/json/BUILD:
```diff
@@ -0,0 +1,21 @@
+        "//src/v/bytes:iobuf",
+        "//src/v/container:fragmented_vector",
+        "//src/v/serde/json/detail:numeric",
+        "//src/v/serde/json/detail:string",
```
these can be implementation_deps

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+
+        case state::first_decimal:
+            if (buf[pos] == '0') {
+                _state = state::decimal_point;
```
Maybe a comment we only accept a single leading 0 for fractional values and otherwise do not support leading zeros?

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+
+        case state::decimals:
+            if (buf[pos] >= '0' && buf[pos] <= '9') {
+                // int64 max/10 = 922337203685477580 but 7
```
I don't understand the "but 7"

wdberkeley replied:
The maximum signed 64-bit integer is `i = 9,223,372,036,854,775,807`, so `i / 10` is `9,223,372,036,854,775,80`. The remainder of 7 is the `but 7`. The code adjusts for this at L68 below. It also handles the minimum signed 64-bit integer value `-9,223,372,036,854,775,808` with the `is_negative` addition.

wdberkeley replied:
I'll adjust the comments to explain this.

rockwotj replied:
Something I had talked with @nvartolomei about was just having this being the state machine for recognizing characters, and linearizing the number into some array that was 32 or 64 bytes or something (longer numbers cause errors). The API would just return a thin wrapper around the string type, with helper methods to parse as double or int64 or uint64 (this is how golang does it with json.Number if you are familiar). Upsides is you don't have to check both number types anymore, and callers already are expecting a type (based on the json schema) and I think it will simplify the caller and impl

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+                    // We are going to overflow the int64_t value.
+                    // Switch to double.
+                    _double_acc = static_cast<double>(_u64_acc);
+                    _as_double = true;
```
Can this be determined based on the end state instead?

rockwotj commented on src/v/serde/json/README.md:
```diff
@@ -0,0 +1,67 @@
+
+- Numeric precision lower than RapidJSON. While this can be ignored as
+  [RFC 8259] does not mandate precision for numbers (section 6), we are still
+  below what C++ double type can represent.
```
So we don't forget, can we record some of the options and tradeoffs we have around linearizing and delegating parsing to a third party vs more testing of overflow and rounding errors? Basically what we talked about in our meeting today

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+
+                if (
+                  _u64_acc > max_int_value
+                  || (_u64_acc == max_int_value && buf[pos] - '0' > 7 + _is_negative)) {
```
Is it simpler to just detect when we overflow during the multiplication?

Or we use absl::int128 to accumulate the number and check after accumulation of the digit if it's over the limit?

wdberkeley replied:
Maybe, but this code is working and tested so I'll leave it as it is for the moment unless you feel strongly about it.

rockwotj replied:
I would prefer to not parse at all and defer to another library over int128

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+                pos += 1;
+                _state = state::exponent_sign;
+                continue;
+            } else if (buf[pos] >= '0' && buf[pos] <= '9') {
```
btw from a performance perspective it might be interesting to see how a lookup table for all these checks fairs. I think you could have a single lookup table and test all these chars you are looking for

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+            if (buf[pos] == '-') {
+                pos += 1;
+                _exp_negative = true;
+                _max_exp = (_exp_frac + 2147483639) / 10;
```
named constant for 21...?

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+                    pos += 1;
+                }
+
+                _max_exp = 308 - _exp_frac;
```
named constant for max exponent?

rockwotj commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+                err = result::done;
+                _state = state::finished_with_double;
+
+                _double_acc = strtod_normal_precision(
```
Another option to consider (I am not sure if this leads to better or worse precision / performance): parse as an absl::int128 and keep trace of the decimal position as the scale then convert to a float. There is plenty of precision available in a decimal 128 value that it should be good. I had to do something similar here: https://github.com/redpanda-data/connect/blob/0181223dd18e755fabaa25890f5f75a71ad542a1/internal/impl/snowflake/streaming/int128/decimal.go#L284

rockwotj commented on src/v/serde/json/detail/numeric.h:
```diff
@@ -0,0 +1,114 @@
+    /// advance() should be called as long as error is result::need_more_data.
+    /// If result is result::done, the parser is done and int64() or double()
+    /// can be called to get the result.
+    size_t advance(ss::temporary_buffer<char>& buf, result& err);
```
can the buffer be const?

andrwng replied:
Same with the string parser i think

rockwotj commented on src/v/serde/json/detail/string.cc:
```diff
@@ -0,0 +1,269 @@
+
+            if (_unicode_index == 4) {
+                unsigned codepoint = parse_codepoint(_unicode_buffer.data());
+                if (codepoint >= 0xD800 && codepoint <= 0xDFFF) {
```
consider using named constants throughout or links to docs on these values

rockwotj replied:
Actually named constants all over here might be a bit much. Calling out it's clearly utf8/utf16 conversions with a link to more resources is likely a better approach

rockwotj commented on src/v/serde/json/parser.cc:
```diff
@@ -0,0 +1,570 @@
+
+namespace experimental::serde::json {
+
+std::ostream& operator<<(std::ostream& os, token t) {
```
nit: use as_format function instead?

nvartolomei replied:
What is `as_format`? Grepping for `as_format` in our codebase doesn't lead to anything.

rockwotj replied:
https://fmt.dev/11.1/api/#udt


The version we are using only supports as_format for enums, the next major version added support for "regular types too".

rockwotj replied:
It's the simpler less verbose way of making your type formattable without fmt::formatter

rockwotj commented on src/v/serde/json/parser.cc:
```diff
@@ -0,0 +1,570 @@
+            } else {
+                break;
+            }
+            co_await ss::coroutine::maybe_yield();
```
Consider using ssx::async_for helpers that limit checking for suspension

rockwotj commented on src/v/serde/json/parser.cc:
```diff
@@ -0,0 +1,570 @@
+    // JSON document. When we suspend, we push the current position on
+    // the stack. When we resume, we pop the position from the stack and
+    // continue parsing from there.
+    chunked_vector<suspension_point> _suspension_stack{
```
Should be a std::vector with depth checks?

rockwotj commented on src/v/serde/json/parser.cc:
```diff
@@ -0,0 +1,570 @@
+        // Try to parse next byte(s) from the buffer based on the current
+        // suspension point.
+        switch (_suspension_stack.back()) {
+        case suspension_point::json_document:
```
Does stuff top level values that are only leaf values (string, number, bool, null)?

nvartolomei replied:
> The root element can be of any JSON data type. Note that in earlier RFC 4627, only objects or arrays were allowed as root values. Since the update to RFC 7159, the root of a JSON document can be of any JSON data type.

- [ ] Must be changed.

rockwotj reviewed (commented):
First pass. I like it! I have not reviewed the tests yet. I do have lots of thoughts on number parsing (not all need to be addressed but if we don't have time to explore alternatives now, let's write them down) and a few questions.

mmaslankaprv commented on src/v/serde/json/detail/string.cc:
```diff
@@ -159,7 +159,9 @@ size_t string_parser::advance(
             }
 
-            throw std::runtime_error(fmt::format("invalid escape: {}", c));
+            err = result::invalid_json_string;
+            _state = state::finished_with_error;
+            return pos;
```
This part can be amended to the commit introducing string parsing.

andrwng commented on src/v/serde/json/tests/parser_bench.cc:
```diff
@@ -0,0 +1,69 @@
+    perf_tests::stop_measuring_time();
+}
+
+PERF_TEST_C(json_parser_bench, serde) { co_await run_serde_bench(); }
+PERF_TEST_C(json_parser_bench, rapidjson) { co_await run_rapidjson_bench(); }
```
nit: can you include some sample output to the commit message so reviewers and future readers understand what to expect from this benchmark?

andrwng commented on src/v/serde/json/detail/string.cc:
```diff
@@ -0,0 +1,267 @@
+
+namespace {
+
+// RapidJSON escape character table.
```
IMO it'd be helpful for reviewers and future readers to point at some RapidJSON code to point out where these concepts are borrowed from RapidJSON

andrwng commented on src/v/serde/json/detail/string.cc:
```diff
@@ -0,0 +1,267 @@
+        case state::start:
+            // Expect the " (start of string) character.
+            if (buf.empty()) {
+                break;
```
nit: dead code due to the while condition?

andrwng commented on src/v/serde/json/detail/string.cc:
```diff
@@ -0,0 +1,267 @@
+                start = pos;
+                continue;
+            }
+
```
What's the significance of throwing here vs returning an error and setting the state as finished_with_error?

wdberkeley replied:
The function should return and put the parser in an error state on invalid escape sequences like `\y`, not throw.

andrwng commented on src/v/serde/json/detail/numeric.cc:
```diff
@@ -0,0 +1,275 @@
+        case state::fractional_part_first_digit:
+            if (buf[pos] >= '0' && buf[pos] <= '9') {
+                _state = state::fractional_part;
+                [[fallthrough]];
```
Is there a conceptual difference between fallthrough and continue? If they're the same, since fallthrough depends on ordering of the switch statement, should we prefer continue instead to make this not so brittle? Or is the idea that fallthrough helps avoid infinite loops or somesuch?

wdberkeley replied:
Yeah I think `continue` makes sense here.

andrwng commented on src/v/serde/json/README.md:
```diff
@@ -23,7 +23,9 @@ while (co_await p.next()) {
 
 ## Known issues
```
```
serde/json: numeric parsing routine

evert "rev"

This reverts commit 60dc774e589f206d897a94fe1bc580d2719d2b2a.
```

nit: extra revert messages

Also here and for other commits, it'd be very much appreciated to mention in the commit message how the commit is expected to fit into the broader JSON parser. With a bit of staring at code it's not hard to infer, but it definitely helps review meaty PRs to include that context up front.

andrwng commented on src/v/serde/json/parser.cc:
```diff
@@ -0,0 +1,570 @@
+        // object `{` or array `[`.
+        json_document,
+
+        // The parser is expecting the first key of an object.
```
nit: of an array?

andrwng commented on src/v/serde/json/parser.cc:
```diff
@@ -0,0 +1,570 @@
+        }
+        TRACE("parse_number, peek {}\n", _buf.peek());
+
+        _numeric_parser = detail::numeric_parser();
```
Is there a particular reason these are members if they're only used here?

andrwng commented on src/v/serde/json/tests/parser_test.cc:
```diff
@@ -0,0 +1,12 @@
+ * by the Apache License, Version 2.0
+ */
+
+#include <gtest/gtest.h>
```
Unintentional include?

EDIT i see it in the next commit. Move it over?

andrwng commented on src/v/serde/json/parser.h:
```diff
@@ -0,0 +1,73 @@
+    token token() const;
+
+    /// Return the current value of the parser.
+    /// Can be called only if a previous call to next_token() returned
```
nit: should refer to just `token()`?

andrwng reviewed (commented):
Nice work! Sorry for the delay in review. IMHO each of the non-test commits could've been their own PR, especially given they have their own comprehensive tests. Though I can understand wanting to get to the final JSON parser to show it working, especially if you're primarily focused on building further on top of this PR

Also, I recall you mentioning you're hoping to merge this to unblock other work. Was the plan to fast-follow these with a redo/cleanup of these parsers? Regardless, some context would be helpful for reviewers in the CL

Generally I agree with others on a couple points, though I don't think it should block if you're planning on cleaning up later (though we should track the cleanup effort):
- the number parsing seems like it could be simplified a bit by using int128, at least for large numbers
- this seems like a great candidate for some fuzzing, though maybe that will be more feasible with subsequent work

wdberkeley commented:
I'm taking over responsibility for this PR. We'd like to merge it soon so we can build some Iceberg with JSON schema capabilities on top of it. I've addressed much of the review feedback, and I think the updated PR is merge-worthy, but there are a few other issues that could be addressed. I'll outline them:
1. Error handling: The parser does not provide good error messages. It just fails. It's important to improve the error handling so we provide at least a description and position for parsing errors in all cases.
2. UTF-8 validation: We do not validate that the text is validly UTF-8 encoded. This is important, and should be straightforward to add.
3. Parse any type at top-level: The updated JSON standard allows any type at the top level, but we accept only objects and arrays. In practice top-level values will be arrays or objects but we should try to match the standard.
4. Depth limits: We do not bail on deeply nested structures. It's important to add this for safety.
5. Number handling: There's a couple of alternate proposals for how to handle number parsing. It'd be nice to simplify, and the current approach has some known issues with high-precision parsing, but the current scheme works pretty well.
6. Fuzzing or randomized testing: We should build some randomized testing, at least. Ideally, we would fuzz test the parser.

wdberkeley committed:
serde/json: Add UTF-8 validation

This adds validation that the input bytes are UTF-8, including checks
that UTF-8 sequences are well-formed, encodings are not overlong, and
codepoints lie in the valid range.

wdberkeley committed:
serde/json: Implement depth limits for nested structures

This implements depth limits for parsing. The limit is configurable when
constructing a parser, with a liberal default limit of 1000 levels.

rockwotj reviewed (approved):
LGTM

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/67813-->
<!--cc425664-df7a-4823-b3f1-1c9cb8931513-->
<!--title-->
## Retry command for [Build#67813](https://buildkite.com/redpanda/redpanda/builds/67813)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}}
```
<!--/body-->

wdberkeley commented:
/ci-repeat 1
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}}

rockwotj commented:
Tests all pass, but the test results job is borked, going to force merge.

dotnwat commented:
@rockwotj @wdberkeley snagged these coverage reports and thought you might be interested in them.

<img width="1105" alt="image" src="https://github.com/user-attachments/assets/a76c9eb0-169d-45c1-84e6-dff8c016f7db" />

<img width="1105" alt="image" src="https://github.com/user-attachments/assets/f6ed0b2f-0418-481f-9e80-018e180c4f7b" />
